edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/Handlers/TwinHandler.cs

```diff
@@ -0,0 +1,350 @@
+                                            Events.ErrorProcessingNotification(e);
+                                        }
+                                    }
+
+                                    Events.ProcessingLoopStopped();
+                                });
+
+            return loopTask;
+        }
+
+        static string GetTwinResultTopic(IIdentity identity, string statusCode, string correlationId)
+        {
+            var identityComponents = identity.Id.Split(HandlerUtils.IdentitySegmentSeparator, StringSplitOptions.RemoveEmptyEntries);
+
+            switch (identityComponents.Length)
+            {
+                case 1: return string.Format(TwinResultDevice, identityComponents[0], statusCode, correlationId);
+                case 2: return string.Format(TwinResultModule, identityComponents[0], identityComponents[1], statusCode, correlationId);
+
+                default:
+                    Events.BadIdentityFormat(identity.Id);
+                    throw new Exception($"cannot decode identity {identity.Id}");
+            }
+        }
+
+        static string GetDesiredPropertiesUpdateTopic(IIdentity identity, string version)
+        {
+            var identityComponents = identity.Id.Split(HandlerUtils.IdentitySegmentSeparator, StringSplitOptions.RemoveEmptyEntries);
+
+            switch (identityComponents.Length)
+            {
+                case 1: return string.Format(DesiredUpdateDevice, identityComponents[0], version);
+                case 2: return string.Format(DesiredUpdateModule, identityComponents[0], identityComponents[1], version);
+
+                default:
+                    Events.BadIdentityFormat(identity.Id);
+                    throw new Exception($"cannot decode identity {identity.Id}");
+            }
+        }
+
+        static class Events
+        {
+            const int IdStart = MqttBridgeEventIds.TwinHandler;
+            static readonly ILogger Log = Logger.Factory.CreateLogger<TwinHandler>();
+
+            enum EventIds
+            {
+                TwinUpdate = IdStart,
+                TwinUpdateFailed,
+                TwinUpdateIncompete,
+                DesiredPropertiesUpdate,
+                DesiredPropertiesUpdateFailed,
+                DesiredPropertiesUpdateIncompete,
+                MissingProxy,
+                UnexpectedTwinTopic,
+                BadIdentityFormat,
+                FailedToSendTwinUpdateMessage,
+                FailedToSendDesiredPropertiesUpdateMessage,
+                BadPayloadFormat,
+                ProcessingLoopStarted,
+                ProcessingLoopStopped,
+                ErrorEnqueueingNotification,
+                ErrorProcessingNotification
+            }
+
+            public static void TwinUpdate(string id, string statusCode, string correlationId, int messageLen) => Log.LogDebug((int)EventIds.TwinUpdate, $"Twin Update sent to client: [{id}], status: [{statusCode}], rid: [{correlationId}], msg len: [{messageLen}]");
```

Comment:
Nit: We don't use the [] pattern elsewhere. I feel coming up with a new pattern for logs will make the EdgeHub logs look very inconsistent.

Reply:
removed, however it is a great pattern - important bits catch the eyes and easier to search (e.g. for [12] vs 12 where the latter stops on every time stamp :) )